internal/keyvalue/leveldb.go

```diff
@@ -83,102 +97,116 @@ func NewLDBDatabase(file string, cache int, handles int) (*LDBDatabase, error) {
 	if _, corrupted := err.(*errors.ErrCorrupted); corrupted {
 		db, err = leveldb.RecoverFile(file, nil)
 	}
-	// (Re)check for errors and abort if opening of the db failed
 	if err != nil {
 		return nil, err
 	}
-	return &LDBDatabase{
-		fn:  file,
-		db:  db,
-		log: logger,
-	}, nil
+	// Assemble the wrapper with all the registered metrics
+	ldb := &LeveldbDatabase{
+		fn:       file,
+		db:       db,
+		log:      logger,
+		quitChan: make(chan chan error),
+	}
+	ldb.compTimeMeter = metrics.NewRegisteredMeter(namespace+"compact/time", nil)
+	ldb.compReadMeter = metrics.NewRegisteredMeter(namespace+"compact/input", nil)
+	ldb.compWriteMeter = metrics.NewRegisteredMeter(namespace+"compact/output", nil)
+	ldb.diskReadMeter = metrics.NewRegisteredMeter(namespace+"disk/read", nil)
+	ldb.diskWriteMeter = metrics.NewRegisteredMeter(namespace+"disk/write", nil)
+	ldb.writeDelayMeter = metrics.NewRegisteredMeter(namespace+"compact/writedelay/duration", nil)
+	ldb.writeDelayNMeter = metrics.NewRegisteredMeter(namespace+"compact/writedelay/counter", nil)
+
+	// Start up the metrics gathering and return
+	go ldb.meter(metricsGatheringInterval)
+	return ldb, nil
 }
 
-// Path returns the path to the database directory.
-func (db *LDBDatabase) Path() string {
-	return db.fn
-}
+// Close stops th emetrics collection, flushes any pending data to disk and closes
+// all io access to the underlying key-value store.
+func (db *LeveldbDatabase) Close() error {
+	db.quitLock.Lock()
+	defer db.quitLock.Unlock()
 
-// Put puts the given key / value to the queue
-func (db *LDBDatabase) Put(key []byte, value []byte) error {
-	return db.db.Put(key, value, nil)
+	if db.quitChan != nil {
+		errc := make(chan error)
+		db.quitChan <- errc
+		if err := <-errc; err != nil {
+			db.log.Error("Metrics collection failed", "err", err)
+		}
+		db.quitChan = nil
+	}
+	return db.db.Close()
 }
 
-func (db *LDBDatabase) Has(key []byte) (bool, error) {
+// Has retrieves if a key is present in the key-value store.
+func (db *LeveldbDatabase) Has(key []byte) (bool, error) {
 	return db.db.Has(key, nil)
 }
 
-// Get returns the given key if it's present.
-func (db *LDBDatabase) Get(key []byte) ([]byte, error) {
+// Get retrieves the given key if it's present in the key-value store.
```

Comment:
Nit: move "in the key-value store" immediately after "the given key"

Reply:
I think it's fine, "if it's present in the key-value store"